river/LayoutManager.zig

```diff
@@ -0,0 +1,378 @@
+                if (self.output.layout_demand.?.serial != req.serial) return;
+
+                // From now on, no matter what happens, we want to destroy the layout demand
+                // and start a transaction.
+                defer {
+                    self.output.layout_demand.?.deinit();
+                    util.gpa.destroy(self.output.layout_demand.?);
+                    self.output.layout_demand = null;
+                    self.output.root.startTransaction();
+                }
+
+                // Check if amount of proposed dimensions matches amount of views to arrange
+                if (self.output.layout_demand.?.views != self.output.layout_demand.?.view_dimensions.len) {
+                    log.err(.layout_manager, "amount of proposed view dimensions does not match amount of views; Aborting layout demand", .{});
+                    layout.postError(
+                        .proposed_dimension_mismatch,
+                        "the amount of proposed view dimensions needs to match the amount of views",
+                    );
+                    return;
+                }
+
+                // Apply proposed layout to views and start transaction
+                var it = ViewStack(View).iter(self.output.views.first, .forward, self.output.pending.tags, Output.arrangeFilter);
+                while (it.next()) |view| {
+                    view.pending.box = self.output.layout_demand.?.popViewDimensions() orelse unreachable;
+                    view.applyConstraints();
+                }
+                self.output.pending.layout = true;
+
+                // TODO layout name?
+            },
+        }
+    }
+};
+
+/// This object encapsulates a zriver.LayoutManagerV1 object bound by a client
+const LayoutManager = struct {
+    parent: *Self,
+    layout_manager: *zriver.LayoutManagerV1,
+
+    /// Initialize a connection
+    pub fn init(self: *LayoutManager, parent: *Self, client: *wl.Client, version: u32, id: u32) !void {
+        self.layout_manager = zriver.LayoutManagerV1.create(client, version, id) catch {
+            self.layout_manager.postNoMemory();
+            log.crit(.river_layout, "out of memory", .{});
+            return Error.BindFailed;
+        };
+        self.parent = parent;
+        self.layout_manager.setHandler(*LayoutManager, handleRequest, null, self);
+    }
+
+    pub fn deinit(self: *LayoutManager) void {
+        self.layout_manager.destroy();
+    }
+
+    fn handleRequest(layout_manager: *zriver.LayoutManagerV1, request: zriver.LayoutManagerV1.Request, self: *LayoutManager) void {
+        switch (request) {
+            .destroy => {
+                self.deinit();
+
+                // Remove this client from the global clients list
+                const node = @fieldParentPtr(std.TailQueue(LayoutManager).Node, "data", self);
+                self.parent.layout_managers.remove(node);
+                util.gpa.destroy(node);
+            },
```

Comment:
Same as above, this should just call `layout_manager.destroy()` and free the memory in the destroy handler set with `setHandler()`